chain/rosetta-rpc/src/adapters/nep141.rs

```diff
@@ -0,0 +1,209 @@
+use std::{collections::HashMap, str::FromStr};
+
+use near_o11y::WithSpanContextExt;
+use near_primitives::{types::BlockId, views::ExecutionOutcomeWithIdView};
+
+use crate::models::{AccountIdentifier, Currency, FungibleTokenEvent};
+
+pub(crate) async fn collect_nep141_events(
+    receipt_execution_outcomes: &Vec<ExecutionOutcomeWithIdView>,
+    block_header: &near_primitives::views::BlockHeaderView,
+    currencies: &Option<Vec<Currency>>,
+) -> crate::errors::Result<Vec<FungibleTokenEvent>> {
+    let mut res = Vec::new();
+    for outcome in receipt_execution_outcomes {
+        let events = extract_events(outcome);
+        for event in events {
+            res.extend(
+                compose_rosetta_nep141_events(&event, outcome, block_header, currencies).await?,
+            );
+        }
+    }
+    Ok(res)
+}
+
+async fn compose_rosetta_nep141_events(
+    events: &crate::models::Nep141Event,
+    outcome: &ExecutionOutcomeWithIdView,
+    block_header: &near_primitives::views::BlockHeaderView,
+    currencies: &Option<Vec<Currency>>,
+) -> crate::errors::Result<Vec<FungibleTokenEvent>> {
+    let mut ft_events = Vec::new();
+    match &events.event_kind {
+        crate::models::Nep141EventKind::FtTransfer(transfer_events) => {
+            if let Some(currencies) = currencies {
+                let currency_map: std::collections::HashMap<String, Currency> =
+                    currencies.clone().into_iter().collect::<HashMap<String, Currency>>();
+                for transfer_event in transfer_events {
+                    if let Some(currency) =
+                        currency_map.get(&outcome.outcome.executor_id.to_string())
+                    {
+                        let base = get_base(Event::Nep141, outcome, block_header)?;
+                        let custom = crate::models::FtEvent {
+                            affected_id: AccountIdentifier::from_str(&transfer_event.old_owner_id)?,
+                            involved_id: Some(AccountIdentifier::from_str(
+                                &transfer_event.new_owner_id,
+                            )?),
+                            delta: -transfer_event.amount.parse::<i64>()?,
```

Comment:
For the parsing of up to 128-bit unsigned integer, this needs special handling. You can probably use the existing `impl DecType for u128` in `core/primitives-core/src/serialize.rs`.

But it won't work with negative values, you should use SignedDiff or something like that.